client/src/reducers/centroidLabels.js

```diff
@@ -0,0 +1,77 @@
+import calcCentroid from "../util/centroid";
+
+const initialState = {
+  labels: [],
+  toggle: false
+};
+
+const centroidLabels = (state = initialState, action, sharedNextState) => {
+  const {
+    world,
+    layoutChoice,
+    categoricalSelection,
+    colors: { colorAccessor }
+  } = sharedNextState;
+
+  switch (action.type) {
+    case "annotation: label current cell selection":
+    case "annotation: label edited":
+    case "annotation: delete label":
+    case "set layout choice":
+    case "set World to current selection":
+    case "reset World to eq Universe":
+      return {
+        ...state,
+        labels:
+          !!colorAccessor &&
+          state.toggle &&
+          !!categoricalSelection[colorAccessor]
+            ? calcCentroid(
+                world.obsAnnotations,
+                world.obsLayout,
+                colorAccessor,
+                layoutChoice.currentDimNames,
+                categoricalSelection,
+                world.schema.annotations.obsByName
+              )
+            : []
+      };
+
+    case "color by categorical metadata":
+    case "show centroid labels for category":
+      if (
+        !colorAccessor ||
+        !(action.toggle || (action.toggle === undefined && state.toggle))
```

Comment:
these conditionals are quite confusing.   At a minimum, I would comment them.  Combining the two cases is clever, but it will be error prone for future updaters (I had to really think about what you were doing)...

Reply:
Resolved using suggested coalesced variable